email/message/src/main/java/com/blazebit/notify/email/message/EmailNotificationMessageResolver.java

```diff
@@ -208,22 +228,24 @@ private static Function<Locale, ResourceBundle> resourceBundleByName(String name
         return locale -> ResourceBundle.getBundle(name, locale);
     }
 
-    private static TemplateProcessor<String> templateProcessorByName(NotificationJobContext jobContext, TemplateContext templateContext, TemplateProcessorFactory<String> templateProcessorFactory, ConfigurationSource configurationSource, String templateName) {
-        if (templateName == null) {
+    private static <T> TemplateProcessor<T> templateProcessorByKey(Class<T> clazz, TemplateContext templateContext, String templateProcessorFactoryName, ConfigurationSource configurationSource, String templateName, ServiceProvider serviceProvider) {
+        if (templateProcessorFactoryName == null) {
             return null;
         }
         if (templateContext == null) {
             throw new NotificationException("No template context given!");
         }
+        TemplateProcessorFactory<T> templateProcessorFactory = templateContext.getTemplateProcessorFactory(templateProcessorFactoryName, clazz);
         if (templateProcessorFactory == null) {
             throw new NotificationException("No template processor factory given!");
         }
         return templateProcessorFactory.createTemplateProcessor(templateContext, key -> {
             if (TemplateProcessor.TEMPLATE_NAME_PROPERTY.equals(key)) {
                 return templateName;
+            } else if (TemplateProcessor.SERVICE_PROVIDER_PROPERTY.equals(key)) {
+                return serviceProvider;
             }
```

Comment:
These two things parameters should be part of the `TemplateProcessorFactory#createTemplateProcessor` method so that we don't have to construct a lambda every time.

Reply:
The EmailNotificationMessageResolver is cached in `com.blazebit.notify.NotificationJobContext.Builder.DefaultNotificationJobContext#getNotificationMessageResolver(java.lang.Class<T>, com.blazebit.job.ConfigurationSource)` so this code should not be executed every time.
Also, how do you suggest to pass these data items into the TemplateProcessorFactory?

Reply:
Just change the `createTemplateProcessor` method to accept these two additional parameters, then you don't need the lambda anymore.